README.md
## Countdown Numbers Solver

  
<!-- PROJECT SHIELDS -->
<!-- Shields are generated dynamically using shield.io with static parameters
URL endpoint is https://img.shields.io/badge/<LABEL>-<MESSAGE>-<COLOR>
Emoji characters are also supported as <LABEL> or <MESSAGE>
Go to https://apps.timwhitlock.info/emoji/tables/unicode and find Bytes (UTF-8) for your emoji, replace "\x" by "%" and use as <LABEL> or <MESSAGE>
Example for smiling face : (Bytes UTF-8) \xF0\x9F\x98\x83 > (<MESSAGE>) %F0%9F%98%83 
-->
[![license: GPLv3](https://img.shields.io/badge/licence-GPLv3-orange)](https://www.gnu.org/licenses/gpl-3.0)
[![python: 3.0](https://img.shields.io/badge/python-3.0-blue)](https://www.python.org/)
[![release version](https://img.shields.io/badge/release-v0.4.2-green)](https://github.com/ingranys/countdown-numbers-solver)
[![maintened](https://img.shields.io/badge/maintened-%E2%9C%94-green)](https://github.com/ingranys/countdown-numbers-solver/commits/)

A Python implementation of a solver for [Countdown Numbers Game, ](https://en.wikipedia.org/wiki/Countdown_(game_show)#Numbers_round) including a wide variety of options.

> **NOTE** : This project is NOT an implementation of Countdown Numbers game. \
> We provide a solver to find all solutions that can hit a given target using the numbers you supplied.

As a reminder here is a **game demo**:

<p align="center">
  <img src="https://github.com/ingranys/countdown-numbers-solver/blob/main/visuals/countdown.gif" height="350">
</p>

Want to have fun? [:point_right:Play the game:point_left:](https://incoherency.co.uk/countdown/practice/#numbers)

Want to get a closer look at the rules and approaches to solve? [:point_right:See blog article:point_left:](https://datagenetics.com/blog/august32014/index.html)


<!-- TABLE OF CONTENTS -->
<!-- Generated automatically (update when saving) using VS Code extension Markdown All in One
-- See https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one 
-->
## Table of Contents
- [Countdown Numbers Solver](#countdown-numbers-solver)
- [Table of Contents](#table-of-contents)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Installation](#installation)
  - [Examples](#examples)
- [Usage](#usage)
- [How It Works](#how-it-works)
  - [The algorithm](#the-algorithm)
  - [Filtering](#filtering)
  - [Complexity](#complexity)
  - [Optimisations](#optimisations)
- [Documentation](#documentation)
  - [Changelog](#changelog)
  - [Comments](#comments)
  - [Docstrings](#docstrings)
- [License :scroll:](#license-scroll)
- [Contact :handshake:](#contact-handshake)
- [Acknowledgements :pray:](#acknowledgements-pray)

<!-- GETTING STARTED -->
## Getting Started

Get a local copy up and run main script.

### Prerequisites
* python3 [:link:](https://www.python.org/downloads/)
* sympy [:link:](https://docs.sympy.org/latest/index.html)
  ```sh
  pip install sympy
  ```

### Installation
1. Clone the repo
   ```sh
   git clone https://github.com/ingranys/countdown-numbers-solver.git
   ```
2. Make sure python3 is installed
   ```sh
   python3 --version
   ```
3. Make sure sympy is installed (see above)
   ```sh
   python3 -c 'import sympy; print("sympy is installed")'
   ```

### Examples
Run script **with numbers and a target**.
```
$ python3 countdown.py -n 100 25 10 3 2 1 -t 784
```
<details>
  <summary>:heavy_plus_sign: See output</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : [100, 25, 10, 3, 2, 1].
  The first 6 number(s) will be considered.
  The goal is to reach 784 using some numbers.
  We are looking for exact solutions only.
  We will filter duplicates and display unique solutions (the order of the operations is ignored).

  >>> START NUMBERS [100, 25, 10, 3, 2, 1]

  >>> TARGET 784

  >>> UNIQUE EXACT SOLUTIONS
  ((100 - 2) * (((10 + 1) * 3) - 25)) = 784
  ((100 - (3 - 1)) * (10 - 2)) = 784
  ((100 - 2) * ((25 - 1) / 3)) = 784
  ((100 + ((25 - 1) / 2)) * (10 - 3)) = 784
  ((100 - 2) * ((10 - 3) + 1)) = 784
  ((100 - 2) * (10 - (3 - 1))) = 784
  (((100 - 3) + 1) * (10 - 2)) = 784
  ```
</details><br />

Run script with numbers and a target and **enforce the use of all number**.
```
$ python3 countdown.py -n 100 25 10 3 2 1 -t 784 -u all
```
<details>
  <summary>:heavy_plus_sign: See output</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : [100, 25, 10, 3, 2, 1].
  The first 6 number(s) will be considered.
  The goal is to reach 784 using all numbers.
  We are looking for exact solutions only.
  We will filter duplicates and display unique solutions.

  >>> START NUMBERS [100, 25, 10, 3, 2, 1]

  >>> TARGET 784

  >>> UNIQUE EXACT SOLUTIONS
  ((100 + ((25 - 1) / 2)) * (10 - 3)) = 784
  ((100 - 2) * (((10 + 1) * 3) - 25)) = 784
  ```
</details><br />

Run script **without providing numbers or a target** (solver will _generate random values instead_).
```
$ python3 countdown.py
```
<details>
  <summary>:heavy_plus_sign: See ouput</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : None given (6 numbers will be picked randomly).
  The goal is to reach 559 using some numbers.
  We are looking for exact solutions only.
  We will filter duplicates and display unique solutions (the order of the operations is ignored).

  >>> START NUMBERS [9, 10, 4, 3, 5, 1]

  >>> TARGET 559

  >>> UNIQUE EXACT SOLUTIONS
  (9 + (10 * (((4 * 3) - 1) * 5))) = 559
  ((((9 + 5) * 4) * 10) - 1) = 559
  (((9 - 1) + 5) * ((10 * 4) + 3)) = 559
  (((((9 + 5) * 10) - 1) * 4) + 3) = 559
  ((9 * (((4 * 3) * 5) + 1)) + 10) = 559
  ((9 + 4) * ((10 * (5 - 1)) + 3)) = 559
  ((9 * ((10 * 5) + (4 * 3))) + 1) = 559
  (((((9 + 3) * 5) - 4) * 10) - 1) = 559
  ```
</details><br />

Run script and look for **approximate answers as well**.
```
$ python3 countdown.py -n 100 75 50 25 1 3 -t 823 -s approximate --tolerance 1 -o unique
```
<details>
  <summary>:heavy_plus_sign: See ouput</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : [100, 75, 50, 25, 1, 3].
  The first 6 number(s) will be considered.
  The goal is to reach 823 using some numbers.
  We are looking for exact and approximate solutions (tolerance is 1).
  We will filter duplicates and display unique solutions.

  >>> START NUMBERS [100, 75, 50, 25, 1, 3]

  >>> TARGET 823

  >>> UNIQUE EXACT SOLUTIONS
  None found...

  >>> UNIQUE APPROXIMATE SOLUTIONS
  (((((100 / 25) * 50) + 75) - 1) * 3) = 822
  (((((100 / 25) * 50) + 75) * 3) - 1) = 824
  (((75 * (50 * (25 - 3))) / 100) - 1) = 824
  (((75 * (25 - 3)) / (100 / 50)) - 1) = 824
  (((((100 * 25) - 75) + 50) / 3) - 1) = 824
  ((((100 * 25) - (75 - 50)) / 3) - 1) = 824
  ((((100 * (75 - 50)) - 25) / 3) - 1) = 824
  ```
</details><br />

Run script and enable **DEBUG mode**.
```
$ python3 countdown.py -n 25 100 1 7 1 5 -t 563 -v 1
```
<details>
  <summary>:heavy_plus_sign: See ouput</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : [25, 100, 1, 7, 1, 5].
  The first 6 number(s) will be considered.
  The goal is to reach 563 using some numbers.
  We are looking for exact solutions only.
  We will filter duplicates and display unique solutions.
  DEBUG mode has been enabled.

  ---------DEBUG---------
  Technical details about args :
  numbers [25, 100, 1, 7, 1, 5]
  size 6
  use some
  target 563
  solutions exact
  tolerance 10
  output unique
  verbose 1
  --------/DEBUG---------

  >>> START NUMBERS [25, 100, 1, 7, 1, 5]

  >>> TARGET 563

  >>> UNIQUE EXACT SOLUTIONS
  ---------DEBUG---------
  SOLUTION | SIMPLIFIED EXPRESSION | EXPRESSION TREE
  (((100 - 1) * 7) - ((25 + 1) * 5)) = 563 | -5*(1 + 25) + (100 + (-1)*1)*7 | Add(Mul(Integer(-1), Mul(Integer(5), Add(Integer(1), Integer(25)))), Mul(Integer(7), Add(Integer(100), Mul(Integer(-1), Integer(1)))))
  --------/DEBUG---------
  (((100 - 1) * 7) - ((25 + 1) * 5)) = 563

  ```
</details><br />

Run script on a very simple example and **enable SPAM mode to investigate how the solver works**.
```
$ python3 countdown.py -n 100 6 10 -t 104 -s approximate --tolerance 5 -o all -v 2
```
<details>
  <summary>:heavy_plus_sign: See ouput</summary>
  
  ```
  
  >>> PARAMETERS
  Input number(s) : [100, 6, 10].
  The first 6 number(s) will be considered.
  The goal is to reach 104 using some numbers.
  We are looking for exact and approximate solutions (tolerance is 5).
  We will display all solutions, even duplicates (operations order will differ).
  DEBUG mode has been enabled.
  !!!SPAM mode has been enabled!!!

  ---------DEBUG---------
  Technical details about args :
  numbers [100, 6, 10]
  size 6
  use some
  target 104
  solutions approximate
  tolerance 5
  output all
  verbose 2
  --------/DEBUG---------

  >>> START NUMBERS [100, 6, 10]

  >>> TARGET 104
  ---  APPROXIMATE SOLUTION FOUND : 100 = 100
  ---  Current number(s) [100, 6, 10] corresponding to the following expression(s) [100, 6, 10]
  ---  Apply operation (+) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [106, 10] corresponding to the following expression(s) ['(100 + 6)', 10]
  ------  APPROXIMATE SOLUTION FOUND : (100 + 6) = 106
  ------  Current number(s) [106, 10] corresponding to the following expression(s) ['(100 + 6)', 10]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [116] corresponding to the following expression(s) ['((100 + 6) + 10)']
  ---------  FINAL VALUE is [116] corresponding to this expression ['((100 + 6) + 10)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [96] corresponding to the following expression(s) ['((100 + 6) - 10)']
  ---------  FINAL VALUE is [96] corresponding to this expression ['((100 + 6) - 10)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 - (100 + 6))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [1060] corresponding to the following expression(s) ['((100 + 6) * 10)']
  ---------  FINAL VALUE is [1060] corresponding to this expression ['((100 + 6) * 10)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((100 + 6) / 10)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [106, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 / (100 + 6))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (-) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [94, 10] corresponding to the following expression(s) ['(100 - 6)', 10]
  ------  Current number(s) [94, 10] corresponding to the following expression(s) ['(100 - 6)', 10]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [104] corresponding to the following expression(s) ['((100 - 6) + 10)']
  ---------  EXACT SOLUTION FOUND : ((100 - 6) + 10) = 104
  ---------  FINAL VALUE is [104] corresponding to this expression ['((100 - 6) + 10)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [84] corresponding to the following expression(s) ['((100 - 6) - 10)']
  ---------  FINAL VALUE is [84] corresponding to this expression ['((100 - 6) - 10)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 - (100 - 6))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [940] corresponding to the following expression(s) ['((100 - 6) * 10)']
  ---------  FINAL VALUE is [940] corresponding to this expression ['((100 - 6) * 10)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((100 - 6) / 10)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [94, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 / (100 - 6))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (^-) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [None, 10] corresponding to the following expression(s) ['(6 - 100)', 10]
  ---  STOP. Unauthorized expression...
  ---  Apply operation (*) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [600, 10] corresponding to the following expression(s) ['(100 * 6)', 10]
  ------  Current number(s) [600, 10] corresponding to the following expression(s) ['(100 * 6)', 10]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [610] corresponding to the following expression(s) ['((100 * 6) + 10)']
  ---------  FINAL VALUE is [610] corresponding to this expression ['((100 * 6) + 10)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [590] corresponding to the following expression(s) ['((100 * 6) - 10)']
  ---------  FINAL VALUE is [590] corresponding to this expression ['((100 * 6) - 10)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 - (100 * 6))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [6000] corresponding to the following expression(s) ['((100 * 6) * 10)']
  ---------  FINAL VALUE is [6000] corresponding to this expression ['((100 * 6) * 10)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [60] corresponding to the following expression(s) ['((100 * 6) / 10)']
  ---------  FINAL VALUE is [60] corresponding to this expression ['((100 * 6) / 10)']
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [600, 10]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(10 / (100 * 6))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (/) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [None, 10] corresponding to the following expression(s) ['(100 / 6)', 10]
  ---  STOP. Unauthorized expression...
  ---  Apply operation (^/) to numbers of indexes (i=0,j=1) in [100, 6, 10]
  ---  Number(s) become(s) [None, 10] corresponding to the following expression(s) ['(6 / 100)', 10]
  ---  STOP. Unauthorized expression...
  ---  Apply operation (+) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [110, 6] corresponding to the following expression(s) ['(100 + 10)', 6]
  ------  Current number(s) [110, 6] corresponding to the following expression(s) ['(100 + 10)', 6]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [116] corresponding to the following expression(s) ['((100 + 10) + 6)']
  ---------  FINAL VALUE is [116] corresponding to this expression ['((100 + 10) + 6)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [104] corresponding to the following expression(s) ['((100 + 10) - 6)']
  ---------  EXACT SOLUTION FOUND : ((100 + 10) - 6) = 104
  ---------  FINAL VALUE is [104] corresponding to this expression ['((100 + 10) - 6)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 - (100 + 10))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [660] corresponding to the following expression(s) ['((100 + 10) * 6)']
  ---------  FINAL VALUE is [660] corresponding to this expression ['((100 + 10) * 6)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((100 + 10) / 6)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [110, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 / (100 + 10))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (-) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [90, 6] corresponding to the following expression(s) ['(100 - 10)', 6]
  ------  Current number(s) [90, 6] corresponding to the following expression(s) ['(100 - 10)', 6]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [96] corresponding to the following expression(s) ['((100 - 10) + 6)']
  ---------  FINAL VALUE is [96] corresponding to this expression ['((100 - 10) + 6)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [84] corresponding to the following expression(s) ['((100 - 10) - 6)']
  ---------  FINAL VALUE is [84] corresponding to this expression ['((100 - 10) - 6)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 - (100 - 10))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [540] corresponding to the following expression(s) ['((100 - 10) * 6)']
  ---------  FINAL VALUE is [540] corresponding to this expression ['((100 - 10) * 6)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [15] corresponding to the following expression(s) ['((100 - 10) / 6)']
  ---------  FINAL VALUE is [15] corresponding to this expression ['((100 - 10) / 6)']
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [90, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 / (100 - 10))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (^-) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [None, 6] corresponding to the following expression(s) ['(10 - 100)', 6]
  ---  STOP. Unauthorized expression...
  ---  Apply operation (*) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [1000, 6] corresponding to the following expression(s) ['(100 * 10)', 6]
  ------  Current number(s) [1000, 6] corresponding to the following expression(s) ['(100 * 10)', 6]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [1006] corresponding to the following expression(s) ['((100 * 10) + 6)']
  ---------  FINAL VALUE is [1006] corresponding to this expression ['((100 * 10) + 6)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [994] corresponding to the following expression(s) ['((100 * 10) - 6)']
  ---------  FINAL VALUE is [994] corresponding to this expression ['((100 * 10) - 6)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 - (100 * 10))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [6000] corresponding to the following expression(s) ['((100 * 10) * 6)']
  ---------  FINAL VALUE is [6000] corresponding to this expression ['((100 * 10) * 6)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((100 * 10) / 6)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [1000, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 / (100 * 10))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (/) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [10, 6] corresponding to the following expression(s) ['(100 / 10)', 6]
  ------  Current number(s) [10, 6] corresponding to the following expression(s) ['(100 / 10)', 6]
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [16] corresponding to the following expression(s) ['((100 / 10) + 6)']
  ---------  FINAL VALUE is [16] corresponding to this expression ['((100 / 10) + 6)']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [4] corresponding to the following expression(s) ['((100 / 10) - 6)']
  ---------  FINAL VALUE is [4] corresponding to this expression ['((100 / 10) - 6)']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 - (100 / 10))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [60] corresponding to the following expression(s) ['((100 / 10) * 6)']
  ---------  FINAL VALUE is [60] corresponding to this expression ['((100 / 10) * 6)']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((100 / 10) / 6)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [10, 6]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(6 / (100 / 10))']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (^/) to numbers of indexes (i=0,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [None, 6] corresponding to the following expression(s) ['(10 / 100)', 6]
  ---  STOP. Unauthorized expression...
  ---  Apply operation (+) to numbers of indexes (i=1,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [100, 16] corresponding to the following expression(s) [100, '(6 + 10)']
  ------  APPROXIMATE SOLUTION FOUND : 100 = 100
  ------  Current number(s) [100, 16] corresponding to the following expression(s) [100, '(6 + 10)']
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [116] corresponding to the following expression(s) ['(100 + (6 + 10))']
  ---------  FINAL VALUE is [116] corresponding to this expression ['(100 + (6 + 10))']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [84] corresponding to the following expression(s) ['(100 - (6 + 10))']
  ---------  FINAL VALUE is [84] corresponding to this expression ['(100 - (6 + 10))']
  ------  Apply operation (^-) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((6 + 10) - 100)']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (*) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [1600] corresponding to the following expression(s) ['(100 * (6 + 10))']
  ---------  FINAL VALUE is [1600] corresponding to this expression ['(100 * (6 + 10))']
  ------  Apply operation (/) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['(100 / (6 + 10))']
  ------  STOP. Unauthorized expression...
  ------  Apply operation (^/) to numbers of indexes (i=0,j=1) in [100, 16]
  ------  Number(s) become(s) [None] corresponding to the following expression(s) ['((6 + 10) / 100)']
  ------  STOP. Unauthorized expression...
  ---  Apply operation (-) to numbers of indexes (i=1,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [100, None] corresponding to the following expression(s) [100, '(6 - 10)']
  ---  STOP. Unauthorized expression...
  ---  Apply operation (^-) to numbers of indexes (i=1,j=2) in [100, 6, 10]
  ---  Number(s) become(s) [100, 4] corresponding to the following expression(s) [100, '(10 - 6)']
  ------  APPROXIMATE SOLUTION FOUND : 100 = 100
  ------  Current number(s) [100, 4] corresponding to the following expression(s) [100, '(10 - 6)']
  ------  Apply operation (+) to numbers of indexes (i=0,j=1) in [100, 4]
  ------  Number(s) become(s) [104] corresponding to the following expression(s) ['(100 + (10 - 6))']
  ---------  EXACT SOLUTION FOUND : (100 + (10 - 6)) = 104
  ---------  FINAL VALUE is [104] corresponding to this expression ['(100 + (10 - 6))']
  ------  Apply operation (-) to numbers of indexes (i=0,j=1) in [100, 4]

CHANGELOG.md
## v0.4.2 (2021-01-09)

*  Update solver documentation. [View](https://github.com/ingranys/countdown-numbers-solver/commit/d77cc49b6514c5546c12e11adc653f8614d5d751)


## v0.4.1 (2021-01-09)

*  Update CHANGELOG.md [View](https://github.com/ingranys/countdown-numbers-solver/commit/d759535787ed6d850f6c3ac011bb271840431fb1)
*  Remove unwanted error. [View](https://github.com/ingranys/countdown-numbers-solver/commit/6bf47931f6158df88f23073b1f9df741d640e0a8)


## v0.4.0 (2021-01-08)

*  Handle errors (ValueError) [View](https://github.com/ingranys/countdown-numbers-solver/commit/bed889934a961cf7c85f7b9729ada0f99f3a3b4d)
*  Update solver.py (minor update to spam messages) [View](https://github.com/ingranys/countdown-numbers-solver/commit/797a9a36efaf484a68582f5d1fe3fb1472a36314)
*  Update all files, add meaningful comments (docstring) [View](https://github.com/ingranys/countdown-numbers-solver/commit/309905e19f2639ad5a758b7364e21205dacf5160)
*  Add changelog-builder.sh and add newly generated CHANGELOG.md. [View](https://github.com/ingranys/countdown-numbers-solver/commit/d0d74d6d2d6985f8c7b45fff3df63e75f59f9f71)


## v0.3.0 (2021-01-07)

*  Make --help menu more readable. [View](https://github.com/ingranys/countdown-numbers-solver/commit/134c21b51977c87af09f9975dd19ecf79cc777bb)
*  Add 'None found...' when empty results. [View](https://github.com/ingranys/countdown-numbers-solver/commit/ed4aadcff004855bfdcf750d80abeb63970e4afc)
*  Add debug message for 'all' solutions. [View](https://github.com/ingranys/countdown-numbers-solver/commit/2f7bf40543bf1b1f283c4e1174152b883862200f)
*  Update filter.py and make it much simplier. [View](https://github.com/ingranys/countdown-numbers-solver/commit/14295f27d219e3626aa16a87251403c9bac66f12)


## v0.2.3 (2021-01-07)

*  Update solver.py, fixing major mismatch between operations and symbols. [View](https://github.com/ingranys/countdown-numbers-solver/commit/0e5b8ad0da75ea14e325ce2bb8e7097ef3bf3aa1)


## v0.2.2 (2021-01-07)

*  Update logger.py for missing spaces. [View](https://github.com/ingranys/countdown-numbers-solver/commit/80a01736cfed81f06e5222271f0547af91a6b0ac)


## v0.2.1 (2021-01-07)

*  Making sure approximate solutions exclude exact solutions. [View](https://github.com/ingranys/countdown-numbers-solver/commit/5ac9ae50762bdbc4075d2c516eb0b5ddf108578c)


## v0.2.0 (2021-01-06)

*  Adding advanced logging and comments. [View](https://github.com/ingranys/countdown-numbers-solver/commit/6e0f5d139be98f3b9f70075dec27d9445f593357)




utils/parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" Custom parser for countdown.

For more information, see README.

For usage, use help menu <python3 countdown.py -h>.

Project can be found here <https://github.com/ingranys/countdown-numbers-solver>.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""

__author__ = "ingranys"
__contact__ = "ingranys@protonmail.com"
__copyright__ = "Copyright 2021, Mustapha Gaies, Toulouse (France)"
__date__ = "2021/01/13"
__deprecated__ = False
__email__ =  "ingranys@protonmail.com"
__license__ = "GPLv3"
__maintainer__ = "developer"
__status__ = "Production"
__version__ = "0.4.2"


from argparse import ArgumentParser
from random import randint


def parse():  
    """
    Parse given arguments according to countdown specifications.
    - Defines arguments names.
    - Verify argument type and limitation on number of values.
    - Verify that values are contained in given choices.
    - Set argument to default when no value is given.
    - Provide usage and help menu.

    Returns:
        argparse.Namespace: Object containing parsed arguments.
    """
  
    parser = ArgumentParser(description="Countdown numbers solver.")
    parser.add_argument(
        "-n",
        "--numbers",
        type=int,
        nargs='*',
        default=None,
        choices=[1,2,3,4,5,6,7,8,9,10,25,50,75,100],
        help="numbers to be used to reach target, "\
            "must be in [1|2|3|4|5|6|7|8|9|10|25|50|75|100] "\
            "(default is random)",
        metavar=''
    )
    parser.add_argument(
        "--size",
        type=int,
        nargs='?',
        default=6,
        choices=range(3,10+1),
        help="maximum number or values to be used (default is 6)",
        metavar='3..10'
    )
    parser.add_argument(
        "-u",
        "--use",
        type=str,
        nargs="?",
        choices=['all','some'],
        default='some',
        help="rule for using numbers (default is 'some')",
        metavar = 'all|some'
    )
    parser.add_argument(
        "-t",
        "--target",
        type=int,
        nargs="?",
        choices=range(100, 999+1),
        default=randint(100,999),
        help="target value (default is random)",
        metavar='100..999'
    )
    parser.add_argument(
        "-s",
        "--solutions",
        type=str,
        nargs="?",
        choices=['approximate','exact'],
        default="exact",
        help="type of authorized answers (default is 'exact')",
        metavar='approximate|exact'
    )
    parser.add_argument(
        "--tolerance",
        type=int,
        nargs="?",
        choices=range(1,10+1),
        default=10,
        help="tolerance range for approximate solutions (default is 10)",
        metavar='1..10'
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        nargs="?",
        choices=['all','unique'],
        default="unique",
        help="solutions to output (default is 'unique')",
        metavar='all|unique'
    )
    parser.add_argument(
        "-v",
        "--verbose",
        type=int,
        nargs="?",
        choices=range(0,3),
        default=0,
        help="verbose level for logging (default is 0)",
        metavar='0|1|2'
    )
    args = parser.parse_args()

    return args